File: lib/elixir_lancedb/schema/field.ex
defmodule ElixirLanceDB.Schema.Field do
  @derive Jason.Encoder
  defstruct ~w(name field_type nullable)a

  alias ElixirLanceDB.Schema.FieldTypes.{Utf8, Float, FixedSizeList}

  @type t() :: %__MODULE__{
          name: String.t(),
          field_type: struct(),
          nullable: boolean()
        }

  def utf8(name, nullable \\ false) do
    %__MODULE__{
      name: name,
      field_type: %Utf8{},
      nullable: nullable
    }
  end

  def float(name, precision, nullable \\ false) do
    %__MODULE__{
      name: name,
      field_type: %Float{
        precision: precision
      },
      nullable: nullable
    }
  end

  def vector(name, %__MODULE__{} = child, dimension) do
    %__MODULE__{
      name: name,
      field_type: %FixedSizeList{
        dimension: dimension,
        child: child
      },
      nullable: true
    }
  end
end

File: native/elixir_lancedb/src/rustler_arrow/schema/field.rs
use std::sync::Arc;

use arrow_schema::{DataType as ArrowDataType, Field as ArrowField};
use rustler::NifStruct;

use super::field_type::FieldType;

#[derive(NifStruct, Clone)]
#[module = "ElixirLanceDB.Native.Schema.Field"]
pub struct Field {
    pub name: String,
    pub field_type: FieldType,
    pub nullable: bool,
}

impl From<&Arc<ArrowField>> for Field {
    fn from(value: &Arc<ArrowField>) -> Self {
        let field_type: FieldType = value.data_type().into();
        Field {
            name: value.name().to_string(),
            field_type: field_type,
            nullable: value.is_nullable(),
        }
    }
}

impl Into<ArrowField> for Field {
    fn into(self) -> ArrowField {
        match self.field_type {
            FieldType::Boolean => ArrowField::new(self.name, ArrowDataType::Boolean, self.nullable),
            FieldType::Utf8 => ArrowField::new(self.name, ArrowDataType::Utf8, self.nullable),
            FieldType::Float32 => ArrowField::new(self.name, ArrowDataType::Float32, self.nullable),
            FieldType::Int32 => ArrowField::new(self.name, ArrowDataType::Int32, self.nullable),
            FieldType::Int64 => ArrowField::new(self.name, ArrowDataType::Int64, self.nullable),
            FieldType::List(child) => ArrowField::new(
                self.name,
                ArrowDataType::List(Arc::new(child.into())),
                self.nullable,
            ),
            FieldType::FixedSizeList(child, dimension) => ArrowField::new(
                self.name,
                ArrowDataType::FixedSizeList(Arc::new(child.into()), dimension),
                self.nullable,
            ),
        }
    }
}
